hbase-asyncfs/src/main/java/org/apache/hadoop/hbase/io/asyncfs/monitor/StreamSlowMonitor.java

```diff
@@ -0,0 +1,207 @@
+   */
+  private static final String DATANODE_SLOW_PACKET_PROCESS_TIME_KEY =
+    "hbase.regionserver.async.wal.datanode.slow.packet.process.time.millis";
+  private static final long DEFAULT_DATANODE_SLOW_PACKET_PROCESS_TIME = 6000; //6s in ms
+
+  /**
+   * Configure for the packet flush speed.
+   * e.g. If the configured slow slow packet process time is larger than 1s, then here 0.1kbs means
+   * 100B should be processed in less than 1s.
+   * If the configured slow slow packet process time is larger than 10s, then here 0.1kbs means
+   * 1KB should be processed in less than 10s.
+   */
+  private static final String DATANODE_SLOW_PACKET_FLUSH_SPEED_KEY =
+    "hbase.regionserver.async.wal.datanode.slow.packet.flush.speed.kbs";
+  private static final double DEFAULT_DATANODE_SLOW_PACKET_FLUSH_SPEED = 0.1;
+
+  private final String name;
+  // this is a map of datanodeInfo->queued slow PacketAckData
+  private final LoadingCache<DatanodeInfo, Deque<PacketAckData>> datanodeSlowDataQueue;
+  private final ExcludeDatanodeManager excludeDatanodeManager;
+
+  private int minSlowDetectCount;
+  private long slowDataTtl;
+  private long slowPacketAckMs;
+  private double slowPacketAckSpeed;
+
+  public StreamSlowMonitor(Configuration conf, String name,
+      ExcludeDatanodeManager excludeDatanodeManager) {
+    setConf(conf);
+    this.name = name;
+    this.excludeDatanodeManager = excludeDatanodeManager;
+    this.datanodeSlowDataQueue = CacheBuilder.newBuilder()
+      .maximumSize(conf.getInt(WAL_MAX_EXCLUDE_SLOW_DATANODE_COUNT_KEY,
+        DEFAULT_WAL_MAX_EXCLUDE_SLOW_DATANODE_COUNT))
+      .expireAfterWrite(conf.getLong(WAL_EXCLUDE_DATANODE_TTL_KEY,
+        DEFAULT_WAL_EXCLUDE_DATANODE_TTL), TimeUnit.HOURS)
+      .build(new CacheLoader<DatanodeInfo, Deque<PacketAckData>>() {
+        @Override
+        public Deque<PacketAckData> load(DatanodeInfo key) throws Exception {
+          return new ConcurrentLinkedDeque<>();
+        }
+      });
+    LOG.info("New stream slow monitor {}", this.name);
+  }
+
+  public static StreamSlowMonitor create(Configuration conf, String name) {
+    return new StreamSlowMonitor(conf, name, new ExcludeDatanodeManager(conf));
+  }
+
+  /**
+   * Check if the packet process time shows that the relevant datanode is a slow node.
+   * @param datanodeInfo the datanode that processed the packet
+   * @param packetDataLen the data length of the packet
+   * @param processTimeMs the process time (in ms) of the packet on the datanode,
+   * @param lastAckTimestamp the last acked timestamp of the packet on another datanode
+   * @param unfinished if the packet is unfinished flushed to the datanode replicas
+   */
+  public void checkProcessTimeAndSpeed(DatanodeInfo datanodeInfo, long packetDataLen,
+      long processTimeMs, long lastAckTimestamp, int unfinished) {
+    long current = EnvironmentEdgeManager.currentTime();
+    // A datanode is suspicious slow if it meets one of the following conditions
```

Comment:
The test condition is a bit strange for me...

In general, we should have two different scenarios here

1. For small packet, we just have a simple limit, without considering the size of the packet.
2. For large packet, we will calculate the speed, and check if the speed is too slow.

On the split point of small and large packet, I do not have a clear idea but 100 seems too small...

Reply:
Hi, @Apache9 , sorry for the late reply.
I think for the two scenarios you described, the speed check is valid for small packets, and the process time limit is for check for any packets.
The default value of "hbase.regionserver.async.wal.datanode.slow.packet.process.time.millis" is 6s, that is for any packets without considering the size processed up to 6s represents a slow datanode existing. But for small packets, since it has small length, the process time on the slow datanode is very likely to be less than 6s. As a result, check the speed of processing small packets is valid here.
I think the split point of small and large packets is relevant to the tolerance of the process time. Here 100 bytes and 0.1 default speed means the check of the process time is at least ONE second, actually is [1s,6s]. Any packet processed less than 1s will be passed. And for the default configs here, the speed will only check packets with 100~600 bytes. For packets larger than 600, if the process speed of it is less than 0.1, then the process time of it will be larger than the default config 6s.
If you think 100 is a little bit small,  for example, set it be 200 means the speed check of the process time is [2s,6s].
But I think a 1s process of the little packet is long enough...

Reply:
I though I was replied before but seems not?

Let me explain more.

Usually, the latency for send 1 byte and 100 byte will be almost the same, as it usually equals to a network round trip. But if you consider the speed, 100 byte will 100x faster than 1 byte, but is it really means we are too slow? That's why I mean we do not need speed check for small packet.

And for large packet, speed should be considered. For example, if you want to send 1GB data, then network bandwidth will be a problem, so usually you will spend 10x times if you want to send 1GB data, comparing to 100MB data. That's why I say, speed check is useful for large packet.

Reply:
OK, got it. Thanks, @Apache9 .
Since the MAX_DATA_LEN in FanOutOneBlockAsyncDFSOutput is 16MB, what about set the length checked here be 4MB?

Reply:
4MB maybe too large, I think 64KB - 256KB will be more suitable here. The more difficult thing is how to choose the slow latency and speed limit...

Reply:
How about set the slow latency be equals to the default "hbase.regionserver.wal.roll.on.sync.ms" which is 10s? Then if we choose the large packet be 128KB, the speed should not be less than 128KB/10s=12.8KB/s.

Reply:
@Apache9 , 64KB seems a reasonable value, and I think 20KB/s is a proper min speed, then 64KB should be processed in less than 3.2s.
I have updated the codes, please take a look in your convenience, thanks.